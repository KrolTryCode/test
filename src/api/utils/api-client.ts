import axios, { AxiosError, InternalAxiosRequestConfig } from 'axios';
import i18n from 'i18next';
import queryString from 'query-string';

import { ApiConfig, Api } from '~/api/utils/api-requests';
import { authPath, loginPath } from '~/utils/configuration/routes-paths';
import {
  projectLocalStorageService,
  projectSessionStorageService,
} from '~/utils/localstorage/project-storage/project-storage-instance';
import { ProjectStorageKey } from '~/utils/localstorage/project-storage/project-storage.types';
import { showErrorMessage } from '~/utils/show-error-message';

export interface ErrorResponse {
  code: number;
  message: string;
}

const getStorageService = () => {
  const rememberMe = projectLocalStorageService.get(ProjectStorageKey.RememberMe);
  return rememberMe ? projectLocalStorageService : projectSessionStorageService;
};

const BASE_API_URL = '/api';

const API_CONFIG: ApiConfig = {
  baseURL: BASE_API_URL,
};

const API_CONFIG_SECURE: ApiConfig = {
  baseURL: BASE_API_URL,
  securityWorker: async () => {
    const token = getStorageService().get(ProjectStorageKey.AccessToken);

    if (token) {
      return { headers: { Authorization: `Bearer ${token ?? ''}` } };
    }
  },
  paramsSerializer: params => {
    const keys = Object.keys(params);

    let res: {
      [key: string]: unknown;
    } = {};

    for (const key of keys) {
      if (typeof params[key] === 'object' && !Array.isArray(params[key])) {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        res = { ...res, ...params[key] };
      } else {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        res[key] = params[key];
      }
    }

    return queryString.stringify(res, { arrayFormat: 'none' });
  },
};

export const ApiClient = new Api(API_CONFIG);

export const ApiClientSecured = new Api(API_CONFIG_SECURE);

ApiClientSecured.instance.interceptors.request.use(request => {
  request.headers.set('Accept-Language', i18n.language);
  return request;
});

ApiClient.instance.interceptors.request.use(request => {
  request.headers.set('Accept-Language', i18n.language);
  return request;
});

// https://gist.github.com/Godofbrowser/bf118322301af3fc334437c683887c5f#file-axios-refresh_token-1-js
let isRefreshing = false;
let failedQueue: {
  resolve: (token: string | null) => void;
  reject: (error: AxiosError) => void;
}[] = [];

const processQueue = (error: AxiosError | null, token: string | null = null) => {
  failedQueue.forEach(prom => {
    error ? prom.reject(error) : prom.resolve(token);
  });
  failedQueue = [];
};

// Можно задать кастомные обработчики для каждой ошибки
ApiClientSecured.instance.interceptors.response.use(
  response => response,
  async (error: AxiosError) => {
    const originalRequest = error.config as InternalAxiosRequestConfig<any> & { _retry: boolean };

    if (error.response?.status === 502) {
      showErrorMessage(error, 'ERROR.TECHNICAL_ERROR');
    }
    if (error.response?.status === 401 && !originalRequest._retry) {
      if (error.config?.url === '/api/v1/jwt/refresh') {
        window.location = `/${authPath}/${loginPath}` as unknown as Location;
        return;
      }

      if (window.location.pathname.startsWith(`/${authPath}`)) {
        return Promise.reject(error);
      }

      if (isRefreshing) {
        return new Promise(function (resolve, reject) {
          failedQueue.push({ resolve, reject });
        })
          .then(token => {
            originalRequest.headers['Authorization'] = 'Bearer ' + String(token);
            return axios(originalRequest);
          })
          .catch(err => {
            return Promise.reject(err);
          });
      }
      originalRequest._retry = true;
      isRefreshing = true;

      const storage = getStorageService();
      storage.remove(ProjectStorageKey.AccessToken);
      const refreshToken = storage.get(ProjectStorageKey.RefreshToken);

      return new Promise(function (resolve, reject) {
        ApiClientSecured.tokenV1Controller
          .regenerateAccessToken({ headers: { Authorization: `Bearer ${refreshToken ?? ''}` } })
          .then(({ token }) => {
            storage.set(ProjectStorageKey.AccessToken, token!);
            axios.defaults.headers.common['Authorization'] = 'Bearer ' + token;
            processQueue(null, token);
            if (originalRequest) {
              originalRequest.headers['Authorization'] = 'Bearer ' + token;
              resolve(axios(originalRequest));
            }
          })
          .catch((err: AxiosError) => {
            processQueue(err, null);
            reject(err);
          })
          .finally(() => {
            isRefreshing = false;
          });
      });
    }
    return Promise.reject(error);
  },
);
