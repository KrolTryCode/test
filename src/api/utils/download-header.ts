import { RequestParams } from './api-requests';

export const getDownloadHeaderCSV = (): RequestParams => ({
  headers: {
    accept: 'text/csv',
  },
});
