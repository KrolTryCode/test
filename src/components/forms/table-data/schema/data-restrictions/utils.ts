import {
  ColumnValue,
  Const,
  FloatConst,
  IntConst,
  StringConst,
  StringLengthValue,
} from '~/api/utils/api-requests';

export const isColumnValueType = (value: any) => {
  return '@type' in value && value['@type'] === 'ColumnValue';
};

export const getRightOrLeftValue = (
  currentObject:
    | string
    | object
    | ColumnValue
    | Const
    | StringLengthValue
    | FloatConst
    | IntConst
    | StringConst,
) => {
  let value: object | string;
  if (typeof currentObject === 'object' && 'value' in currentObject) {
    value = currentObject.value;
  } else {
    value = currentObject;
  }
  return value;
};
