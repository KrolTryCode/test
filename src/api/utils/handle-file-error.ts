import { AxiosError } from 'axios';

import { showErrorMessage } from '~/utils/show-error-message';

export function handleFileUploadingError(e: unknown) {
  if (e instanceof AxiosError && e.response?.status === 413) {
    showErrorMessage(e, 'FILES.ERROR.TOO_BIG');
  } else {
    showErrorMessage(e, 'FILES.UPLOAD_FAILED');
  }
}
