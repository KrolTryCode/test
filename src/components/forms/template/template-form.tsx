import { yupResolver } from '@hookform/resolvers/yup';
import { Form, FormItem, FormButtons, Button } from '@pspod/ui-components';
import { FC } from 'react';
import { useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';

import { CreateTemplateRequest } from '~/api/utils/api-requests';
import { ButtonLink } from '~/components/implicit-links';
import { FormInputText, FormInputRichTextEditor } from '~/components/react-hook-form';

import { schema } from './template-form.schema';

interface TemplateFormProps {
  data: CreateTemplateRequest;
  onSave: (data: Required<CreateTemplateRequest>) => void;
  isPending?: boolean;
  isLoading?: boolean;
}

export const TemplateForm: FC<TemplateFormProps> = ({ data, isLoading, isPending, onSave }) => {
  const { t } = useTranslation();

  const {
    register,
    handleSubmit,
    control,
    formState: { isValid, isSubmitted },
  } = useForm<CreateTemplateRequest>({
    mode: 'onBlur',
    reValidateMode: 'onBlur',
    values: data,
    defaultValues: schema.getDefault(),
    resolver: yupResolver(schema),
  });

  return (
    <Form isRequired isLoading={isLoading} onSubmit={handleSubmit(onSave)}>
      <FormItem label={t('COMMON.TITLE')} isDisabled>
        <FormInputText isReadonly controllerProps={{ ...register('name'), control }} />
      </FormItem>
      <FormItem>
        <FormInputRichTextEditor controllerProps={{ ...register('content'), control }} />
      </FormItem>
      <FormButtons marginTop={1}>
        <ButtonLink to={'/admin/templates'} variant={'outlined'} color={'primary'}>
          {t('ACTION.CANCEL')}
        </ButtonLink>
        <Button
          type={'submit'}
          variant={'contained'}
          color={'primary'}
          disabled={!isValid && isSubmitted}
          isLoading={isPending}
        >
          {t('ACTION.SAVE')}
        </Button>
      </FormButtons>
    </Form>
  );
};
