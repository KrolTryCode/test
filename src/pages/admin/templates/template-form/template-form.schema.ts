import * as y from 'yup';

import { CreateTemplateRequest } from '~/api/utils/api-requests';
import { nonEmptyStringYup } from '~/utils/validation/schemas/strings';

export const schema: y.ObjectSchema<CreateTemplateRequest> = y.object().shape({
  name: nonEmptyStringYup,
  content: nonEmptyStringYup,
});

export const defaultValues: CreateTemplateRequest = {
  name: '',
  content: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Тема письма</title>
  </head>
  <body>

  </body>
</html>`,
};
