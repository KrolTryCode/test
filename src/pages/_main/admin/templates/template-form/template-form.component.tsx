import { yupResolver } from '@hookform/resolvers/yup';
import { Form, FormItem, FormButtons, Button, notifySuccess } from '@pspod/ui-components';
import { useNavigate } from '@tanstack/react-router';
import { FC, useCallback } from 'react';
import { useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';

import { useUpdateTemplateMutation } from '~/api/queries/templates/update-template.mutation';
import { CreateTemplateRequest, Template } from '~/api/utils/api-requests';
import { ButtonLink } from '~/components/implicit-links';
import { FormInputText } from '~/components/react-hook-form';
import { FormInputRichTextEditor } from '~/components/react-hook-form/form-input-rich-editor/form-input-rich-editor.component';
import { showErrorMessage } from '~/utils/show-error-message';

import { defaultValues, schema } from './template-form.schema';

export type TemplateFormProps = {
  data?: Template;
};

export const TemplateForm: FC<TemplateFormProps> = ({ data }) => {
  const { t } = useTranslation();
  const navigate = useNavigate();

  const {
    register,
    handleSubmit,
    control,
    formState: { isValid, isSubmitted },
  } = useForm<CreateTemplateRequest>({
    mode: 'onBlur',
    reValidateMode: 'onBlur',
    defaultValues: data ?? defaultValues,
    resolver: yupResolver(schema),
  });

  const updateTemplateMutation = useUpdateTemplateMutation();

  const onSubmit = useCallback(
    async (template: Template) => {
      await updateTemplateMutation.mutateAsync(
        {
          templateId: data?.id ?? '',
          content: template.content ?? '',
        },
        {
          onSuccess: () => {
            notifySuccess(t('MESSAGE.UPDATE_SUCCESS'));
            void navigate({ to: '/admin/templates' });
          },
          onError: e => showErrorMessage(e, 'ERROR.UPDATE_FAILED'),
        },
      );
    },
    [data?.id, navigate, t, updateTemplateMutation],
  );

  return (
    <Form onSubmit={handleSubmit(onSubmit)}>
      <FormItem isRequired label={t('COMMON.TITLE')} isDisabled>
        <FormInputText isReadonly controllerProps={{ ...register('name'), control }} />
      </FormItem>
      <FormItem isRequired>
        <FormInputRichTextEditor controllerProps={{ ...register('content'), control }} />
      </FormItem>
      <FormButtons marginTop={2}>
        <ButtonLink to={'/admin/templates'} variant={'outlined'} color={'primary'}>
          {t('ACTION.CANCEL')}
        </ButtonLink>
        <Button
          type={'submit'}
          variant={'contained'}
          color={'primary'}
          disabled={!isValid && isSubmitted}
          isLoading={updateTemplateMutation.isPending}
        >
          {t('ACTION.SAVE')}
        </Button>
      </FormButtons>
    </Form>
  );
};
