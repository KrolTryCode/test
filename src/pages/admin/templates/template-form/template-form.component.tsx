import { yupResolver } from '@hookform/resolvers/yup';
import { Form, FormItem, FormButtons, Button, notifySuccess } from '@pspod/ui-components';
import { FC, useCallback } from 'react';
import { useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { useNavigate } from 'react-router-dom';

import { useUpdateTemplateMutation } from '~/api/queries/templates/update-template.mutation';
import { CreateTemplateRequest, Template } from '~/api/utils/api-requests';
import { FormInputText } from '~/components/react-hook-form';
import { FormInputRichTextEditor } from '~/components/react-hook-form/form-input-rich-editor/form-input-rich-editor.component';
import { adminPath, templatesPath } from '~/utils/configuration/routes-paths';
import { showErrorMessage } from '~/utils/show-error-message';

import { defaultValues, schema } from './template-form.schema';

export type TemplateFormProps = {
  data?: Template;
};

export const TemplateForm: FC<TemplateFormProps> = ({ data }) => {
  const { t } = useTranslation();
  const navigate = useNavigate();

  const {
    register,
    handleSubmit,
    control,
    formState: { isValid, isSubmitted },
  } = useForm<CreateTemplateRequest>({
    mode: 'onBlur',
    reValidateMode: 'onBlur',
    defaultValues: data ?? defaultValues,
    resolver: yupResolver(schema),
  });

  const updateTemplateMutation = useUpdateTemplateMutation();

  const onSubmit = useCallback(
    async (template: Template) => {
      await updateTemplateMutation.mutateAsync(
        {
          templateId: data?.id ?? '',
          content: template.content ?? '',
        },
        {
          onSuccess: () => {
            notifySuccess(t('MESSAGE.UPDATE_SUCCESS'));
            navigate(`/${adminPath}/${templatesPath}`);
          },
          onError: e => showErrorMessage(e, 'ERROR.UPDATE_FAILED'),
        },
      );
    },
    [data?.id, navigate, t, updateTemplateMutation],
  );

  return (
    <Form onSubmit={handleSubmit(onSubmit)}>
      <FormItem isRequired label={t('COMMON.TITLE')} isDisabled>
        <FormInputText isReadonly controllerProps={{ ...register('name'), control }} />
      </FormItem>
      <FormItem isRequired>
        <FormInputRichTextEditor controllerProps={{ ...register('content'), control }} />
      </FormItem>
      <FormButtons marginTop={2}>
        <Button onClick={() => navigate(-1)} variant={'outlined'} color={'primary'}>
          {t('ACTION.CANCEL')}
        </Button>
        <Button
          type={'submit'}
          variant={'contained'}
          color={'primary'}
          disabled={!isValid && isSubmitted}
          isLoading={updateTemplateMutation.isPending}
        >
          {t('ACTION.SAVE')}
        </Button>
      </FormButtons>
    </Form>
  );
};
