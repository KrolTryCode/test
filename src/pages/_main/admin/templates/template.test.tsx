import { screen } from '@testing-library/react';
import { it, describe, expect } from 'vitest';

import { mockGetTemplates } from 'tests/__mocks__/api-client.mock';
import { renderComponent } from 'tests/utils/render-component';
import { CreateTemplateRequest } from '~/api/utils/api-requests';

import { schema } from './template-form/template-form.schema';

import { TemplatesPage } from '.';

const correctFormData: CreateTemplateRequest = {
  name: 'Template name',
  content:
    '<!doctype html><html lang="en"><head><meta charset="UTF-8"><meta name="viewport" content="width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0"><meta http-equiv="X-UA-Compatible" content="ie=edge"><title>Код реактивации аккаунта</title></head><body><div>Ваш код реактивации: <br><b>${activationCode}</b> <br><a href="${activationUrl}">Или нажмите на эту ссылку</a></div><br/></body></html>',
};

describe('Create template form', () => {
  it('should pass with correct data', async () => {
    const parsed = schema.isValid({ ...correctFormData });
    await expect(parsed).resolves.toBeTruthy();
  });

  it('should fail with empty data', async () => {
    await expect(schema.isValid({})).resolves.toBeFalsy();
    await expect(
      schema.isValid({ ...correctFormData, name: '', content: '' }),
    ).resolves.toBeFalsy();
  });
});

describe.skip('Templates page', () => {
  beforeAll(() => {
    renderComponent(<TemplatesPage />, { withRouter: true });
  });

  it('renders the templates', async () => {
    const dataGrid = await screen.findByRole('grid');
    const columns = await screen.findAllByRole('columnheader');
    expect(dataGrid).toBeInTheDocument();
    expect(columns).toHaveLength(4);
    expect(mockGetTemplates).toHaveBeenCalledTimes(1);

    expect(dataGrid).toMatchSnapshot();
  });
});
