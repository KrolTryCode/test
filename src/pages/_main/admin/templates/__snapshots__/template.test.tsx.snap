// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Templates page > renders the templates 1`] = `
<div
  aria-colcount="4"
  aria-multiselectable="true"
  aria-rowcount="1"
  class="MuiDataGrid-main css-1bgckiu-MuiDataGrid-main"
  role="grid"
  tabindex="-1"
>
  <div
    class="css-18l11qw"
    data-id="gridPanelAnchor"
    role="presentation"
  />
  <div
    class="MuiDataGrid-virtualScroller MuiDataGrid-virtualScroller--hasScrollX css-1vigakz-MuiDataGrid-virtualScroller"
    role="presentation"
    style="overflow-x: hidden;"
  >
    <div
      class="MuiDataGrid-topContainer MuiDataGrid-container--top css-9dbfce"
      role="presentation"
    >
      <div
        class="MuiDataGrid-columnHeaders css-1gtv474-MuiDataGrid-columnHeaders"
        role="presentation"
      >
        <div
          aria-rowindex="1"
          class="MuiDataGrid-row--borderBottom css-yseucu-MuiDataGrid-columnHeaderRow"
          role="row"
        >
          <div
            role="presentation"
            style="width: 0px;"
          />
          <div
            aria-colindex="1"
            aria-sort="none"
            class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--sortable MuiDataGrid-withBorderColor MuiDataGrid-columnHeader--withRightBorder"
            data-field="name"
            role="columnheader"
            style="height: 40px; width: 0px;"
            tabindex="0"
          >
            <div
              class="MuiDataGrid-columnHeaderDraggableContainer"
              draggable="true"
              role="presentation"
            >
              <div
                class="MuiDataGrid-columnHeaderTitleContainer"
                role="presentation"
              >
                <div
                  class="MuiDataGrid-columnHeaderTitleContainerContent"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitle css-s39ini-MuiDataGrid-columnHeaderTitle"
                    data-mui-internal-clone-element="true"
                  >
                    Название
                  </div>
                </div>
                <div
                  class="MuiDataGrid-iconButtonContainer css-ltf0zy-MuiDataGrid-iconButtonContainer"
                >
                  <button
                    aria-label="Сортировать"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1p0j4z7-MuiButtonBase-root-MuiIconButton-root"
                    field="name"
                    tabindex="-1"
                    title="Сортировать"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall MuiDataGrid-sortIcon css-120dh41-MuiSvgIcon-root"
                      data-testid="ArrowUpwardIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                      />
                    </svg>
                  </button>
                </div>
              </div>
              <div
                class="MuiDataGrid-menuIcon"
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="menu"
                  aria-label="Меню"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiDataGrid-menuIconButton css-1p0j4z7-MuiButtonBase-root-MuiIconButton-root"
                  data-mui-internal-clone-element="true"
                  id=":r7:"
                  tabindex="-1"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1ckov0h-MuiSvgIcon-root"
                    data-testid="TripleDotsVerticalIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                    />
                  </svg>
                </button>
              </div>
            </div>
            <div
              class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--resizable MuiDataGrid-columnSeparator--sideRight"
              style="min-height: 40px;"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-1umw9bq-MuiSvgIcon-root"
                data-testid="SeparatorIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <rect
                  height="24"
                  rx="0.5"
                  width="1"
                  x="11.5"
                />
              </svg>
            </div>
          </div>
          <div
            aria-colindex="2"
            aria-sort="none"
            class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--sortable MuiDataGrid-withBorderColor MuiDataGrid-columnHeader--withRightBorder"
            data-field="state"
            role="columnheader"
            style="height: 40px; width: 0px;"
            tabindex="-1"
          >
            <div
              class="MuiDataGrid-columnHeaderDraggableContainer"
              draggable="true"
              role="presentation"
            >
              <div
                class="MuiDataGrid-columnHeaderTitleContainer"
                role="presentation"
              >
                <div
                  class="MuiDataGrid-columnHeaderTitleContainerContent"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitle css-s39ini-MuiDataGrid-columnHeaderTitle"
                    data-mui-internal-clone-element="true"
                  >
                    Состояние
                  </div>
                </div>
                <div
                  class="MuiDataGrid-iconButtonContainer css-ltf0zy-MuiDataGrid-iconButtonContainer"
                >
                  <button
                    aria-label="Сортировать"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1p0j4z7-MuiButtonBase-root-MuiIconButton-root"
                    field="state"
                    tabindex="-1"
                    title="Сортировать"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall MuiDataGrid-sortIcon css-120dh41-MuiSvgIcon-root"
                      data-testid="ArrowUpwardIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                      />
                    </svg>
                  </button>
                </div>
              </div>
              <div
                class="MuiDataGrid-menuIcon"
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="menu"
                  aria-label="Меню"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiDataGrid-menuIconButton css-1p0j4z7-MuiButtonBase-root-MuiIconButton-root"
                  data-mui-internal-clone-element="true"
                  id=":rd:"
                  tabindex="-1"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1ckov0h-MuiSvgIcon-root"
                    data-testid="TripleDotsVerticalIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                    />
                  </svg>
                </button>
              </div>
            </div>
            <div
              class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--resizable MuiDataGrid-columnSeparator--sideRight"
              style="min-height: 40px;"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-1umw9bq-MuiSvgIcon-root"
                data-testid="SeparatorIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <rect
                  height="24"
                  rx="0.5"
                  width="1"
                  x="11.5"
                />
              </svg>
            </div>
          </div>
          <div
            aria-colindex="3"
            aria-sort="none"
            class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--sortable MuiDataGrid-withBorderColor MuiDataGrid-columnHeader--withRightBorder"
            data-field="content"
            role="columnheader"
            style="height: 40px; width: 0px;"
            tabindex="-1"
          >
            <div
              class="MuiDataGrid-columnHeaderDraggableContainer"
              draggable="true"
              role="presentation"
            >
              <div
                class="MuiDataGrid-columnHeaderTitleContainer"
                role="presentation"
              >
                <div
                  class="MuiDataGrid-columnHeaderTitleContainerContent"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitle css-s39ini-MuiDataGrid-columnHeaderTitle"
                    data-mui-internal-clone-element="true"
                  >
                    Содержание
                  </div>
                </div>
                <div
                  class="MuiDataGrid-iconButtonContainer css-ltf0zy-MuiDataGrid-iconButtonContainer"
                >
                  <button
                    aria-label="Сортировать"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1p0j4z7-MuiButtonBase-root-MuiIconButton-root"
                    field="content"
                    tabindex="-1"
                    title="Сортировать"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall MuiDataGrid-sortIcon css-120dh41-MuiSvgIcon-root"
                      data-testid="ArrowUpwardIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                      />
                    </svg>
                  </button>
                </div>
              </div>
              <div
                class="MuiDataGrid-menuIcon"
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="menu"
                  aria-label="Меню"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiDataGrid-menuIconButton css-1p0j4z7-MuiButtonBase-root-MuiIconButton-root"
                  data-mui-internal-clone-element="true"
                  id=":rj:"
                  tabindex="-1"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1ckov0h-MuiSvgIcon-root"
                    data-testid="TripleDotsVerticalIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                    />
                  </svg>
                </button>
              </div>
            </div>
            <div
              class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--resizable MuiDataGrid-columnSeparator--sideRight"
              style="min-height: 40px;"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-1umw9bq-MuiSvgIcon-root"
                data-testid="SeparatorIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <rect
                  height="24"
                  rx="0.5"
                  width="1"
                  x="11.5"
                />
              </svg>
            </div>
          </div>
          <div
            aria-colindex="4"
            aria-sort="none"
            class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--alignCenter MuiDataGrid-withBorderColor MuiDataGrid-columnHeader--lastUnpinned MuiDataGrid-columnHeader--last"
            data-field="actions"
            role="columnheader"
            style="height: 40px; width: 50px;"
            tabindex="-1"
          >
            <div
              class="MuiDataGrid-columnHeaderDraggableContainer"
              draggable="false"
              role="presentation"
            >
              <div
                class="MuiDataGrid-columnHeaderTitleContainer"
                role="presentation"
              >
                <div
                  class="MuiDataGrid-columnHeaderTitleContainerContent"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitle css-s39ini-MuiDataGrid-columnHeaderTitle"
                    data-mui-internal-clone-element="true"
                  />
                </div>
              </div>
            </div>
            <div
              class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
              style="min-height: 40px;"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-1umw9bq-MuiSvgIcon-root"
                data-testid="SeparatorIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <rect
                  height="24"
                  rx="0.5"
                  width="1"
                  x="11.5"
                />
              </svg>
            </div>
          </div>
          <div
            class="MuiDataGrid-filler"
            role="presentation"
          />
          <div
            class="MuiDataGrid-scrollbarFiller MuiDataGrid-scrollbarFiller--header"
            role="presentation"
          />
        </div>
        <div
          class="css-4f60me"
        />
      </div>
    </div>
    <div
      class="MuiDataGrid-overlayWrapper css-74bi4q-MuiDataGrid-overlayWrapper"
    >
      <div
        class="MuiDataGrid-overlayWrapperInner css-1mbsuzk-MuiDataGrid-overlayWrapperInner"
        style="height: -40px; width: 0px;"
      >
        <div
          class="MuiDataGrid-overlay css-ok32b7-MuiDataGrid-overlay"
        >
          Нет строк
        </div>
      </div>
    </div>
    <div
      class="MuiDataGrid-virtualScrollerContent css-aymtem-MuiDataGrid-virtualScrollerContent"
      role="presentation"
      style="width: auto; flex-basis: 0px; flex-shrink: 0;"
    >
      <div
        class="MuiDataGrid-virtualScrollerRenderZone css-11dqcl8-MuiDataGrid-virtualScrollerRenderZone"
        role="rowgroup"
        style="transform: translate3d(0, 0px, 0);"
      />
    </div>
    <div
      class="MuiDataGrid-bottomContainer MuiDataGrid-container--bottom css-kkb6k7"
      role="presentation"
    />
  </div>
  <div
    aria-hidden="true"
    class="MuiDataGrid-scrollbar MuiDataGrid-scrollbar--horizontal css-ad8qk9"
    tabindex="-1"
  >
    <div
      class="MuiDataGrid-scrollbarContent"
      style=""
    />
  </div>
</div>
`;
